### PR TITLE
Features/multi period

### PR DESCRIPTION
**Pull request by jokochems**
[EDITED POST]
This one serves to integrate the multi-period feature into the stable code basis.

It is the successor of https://github.com/oemof/oemof-solph/pull/810

A single period-modeling as we have it is going to become a special case of multi-period modeling.
The feature has been adjusted to the new API structure v0.5 will bring along.

Related issue: closes https://github.com/oemof/oemof-solph/issues/793

Feel free to request any changes.

I have included two example test_scripts as well as a step by step guide in the `usage.rst` file.

Main changes:
* Add `discount_rate` to model.
* Add attribute `periods` to energy system to indicate that a multi-period model shall be considered.
* Create new timesets:
  * PERIODS: used for investment variables
  * TIMEINDEX: tuple of (period, timestep)
* Index flow variable in TIMEINDEX. Adjust every constraint that includes a flow; but do not change indexing of variables indexed in TIMESTEPS. 
* Add additional attributes to _options.Investment and GenericStorage / SinkDSM: `lifetime`, `age`, `interest_rate`, `fixed_costs`.
* Add variables `old`, `old_end` and `old_exo` in investment blocks for lifetime tracking.
* Add adjusted objective value terms for multi-period model including discounting and annuities for investment.
* Adjust / restructure processing to properly retrieve results.

I'm sorry, the diffs are huge. We also could have dedicated webinars / sessions on this one if you'd like.

### CODE REVIEW

**Comment by jokochems** [comment]
Hum, weird. Again quite some black issues popping up after merging it from my fork. I'll tackle them tomorrow. Don't mind if you already plan to take a look.

**Comment by jokochems** [comment]
I updated this Pull Request to the latest version of branch `v0.5`. What is very weird is that the pytest for a newly introduced multi_period_dispatch example fails. It does not fail on my local machine when running a tox session, though.

Unfortunately, at the moment, I am not able to properly debug this, because I am experiencing some issue with solvers being not properly set up and interfering with some misconfiguration in the windows or python settings which I have not been able to fix yet. 

@p-snft or @uvchik: Feel free to take a look in particular at this single failing test if your time allows. I think, it would also be okay to leave this one as it is for now and take care of the failing test later.

**Comment by p-snft** [comment]
Didn't look at it yet, but we once had some failing  constraint tests because of rounding issues. We compare ASCII representations of the resulting formulas, which include non-integers. (It's strongly advised not to compare floating point variables for equality, but it's the best we can do here, imho.)

**Comment by jokochems** [comment]
> Didn't look at it yet, but we once had some failing constraint tests because of rounding issues. We compare ASCII representations of the resulting formulas, which include non-integers. (It's strongly advised not to compare floating point variables for equality, but it's the best we can do here, imho.)

Thank you! I've come across such rounding precision issues (also comparing among solvers). I think, this might not be the case here. I evaluate the cumulated storage flows and round these to an integer in order to compare. It finds a cumulated value of 20 in the optimal objective compared to 30 which should be the correct one. I think, this might relate to the new handling of timesteps in combination with results processing. But I wonder why the tests on the local machine don't fail.

**Comment by p-snft** [comment]
Status update: I was able to reproduce both, the error and the run without the error. Python 3.7 fails but Python 3.8 works as intended. Can you confirm that you are not using Python 3.7 on your local machine?

> tests/test_scripts/test_solph/test_multi_period_model/test_multi_period_dispatch_model.py:308:
.tox/py37/lib/python3.7/site-packages/nose/tools/trivial.py:29: in eq_
    raise AssertionError(msg or "%r != %r" % (a, b))
E   AssertionError: 20 != 30
...
ERROR:   py37: commands failed


But:
>  py38: commands succeeded

**Comment by jokochems** [comment]
Thank you very much for looking into it so quickly, @p-snft 

> Can you confirm that you are not using Python 3.7 on your local machine?

Yes, I can confirm that. I am not using Python 3.7.

That's very interesting that Python 3.8 and 3.9 work as intended while Python 3.7 fails and actually finds another solution.

**Comment by p-snft** [comment]
I had a slightly deeper look: The problem occurs when the key in the last lines of `test_multi_period_dispatch_model` is `DE_storage_el`. For me, it looks as if the readout of the `storage_content` was broken:
grafik

**Comment by jokochems** [comment]
Here's a memo on the near-future TODOs for myself (and also for you to follow further dev):

- [ ] Resolve merge conflicts (e.g. docs #857) and update to new flow structure (#856)
- [ ] Update SinkDSM's documentation (`flex_share_up` and `flex_share_down` are dropped)
- [ ] Alter `EnergySystem` to allow for different frequencies

**Comment by p-snft** [comment]
> Dig into failing test and (weird) python 3.7 behaviour; check numpy, pandas, ... versions and find out whether default behaviour changed

Good news: That's a "wontfix", as we [dropped support for Python 3.7](https://github.com/oemof/oemof-solph/commit/fc37637e57362142068f66d3de7c48645740f9ca).

**Comment by jokochems** [comment]
I updated this Pull Request and removed the merge conflicts. Luckily, it was all possible, though requiring around 1.5 - 2 working days of effort.

@p-snft:
* Obviously, the `pytest-travis-fold` dependency listed in the `tox.ini` file caused some trouble and led to an (import) Error which I could not make much sense of. - I'm sorry, I'm not so very familiar with tox and how it works.
* I removed this dependency, but I'm not sure if I can safely do so.
* Can I safely remove the dependency (since Travis CI isn't used anyways anymore) and if not, how should I proceed to resolve this issue?
Thank you very much! 😃

**Comment by jokochems** [comment]
Since there are not any conflicts anymore and the CI pipeline succeeds, I'll mark this pull request as ready, indicating that it is ready to go into the code basis and the further step to be taken within this month are to be discussed at the upcoming dev meeting (see https://github.com/oemof/oemof/issues/100 resp. https://github.com/oemof/oemof/wiki/Meeting-2022.11).

@oemof-developer and esp. @p-snft and @uvchik: It would be awesome if you were able to have a glimpse in advance of the dev meeting next week, but I do not (yet 😉) expect you to go through it change by change. Thank you!

**Review by p-snft — COMMENTED** [review]
Review is not really complete.

File: docs/whatsnew/v0-5-0.rst [hunk 1]
```diff
@@ -14,7 +14,7 @@ API changes
 New features
 ############
 
-* Add `inactivity_costs` as an option for `Flow`s.
+* Add `inactivity_costs` as an option for `Flow`\s.
```
[the remark this answers is not in the document]
Comment (jokochems) [reply docs/whatsnew/v0-5-0.rst:17]: Yes, it is intentional. I added it since otherwise it led to a sphinx warning which was interpreted as an error and thus led to a failing docs test.

File: src/oemof/solph/_energy_system.py [hunk 2]
```diff
@@ -47,6 +49,21 @@ class EnergySystem(es.EnergySystem):
...
+
+    multi_period : boolean
+        If True, a multi-period model is used; defaults to False
+
+    periods : dict
+        The periods of a multi-period model;
+        Keys are the numbers of the respective period, starting with zero,
+        values are pd.date_range objects carrying the timeindex for the
+        respective period;
+        For a standard model, periods only holds one entry, i.e. {0: 0}
```
[the remark this answers is not in the document]
Comment (jokochems) [reply src/oemof/solph/_energy_system.py:61]: Yes, you are right: One could also use list indices. But if you want to explicitly define periods, a dict might be a bit handier to use.

File: src/oemof/solph/_energy_system.py [hunk 3]
```diff
@@ -107,12 +127,20 @@ def __init__(
...
+                    "conflicting to each other."
+                )
+                raise AttributeError(msg)
+            else:
+                msg = (
+                    "Ensure that your timeindex and timeincrement are "
+                    "consistent.\nIf you are not considering non-equidistant "
+                    "timeindices, consider only specifying a timeindex."
+                )
+                warnings.warn(msg, debugging.SuspiciousUsageWarning)
```
[the remark this answers is not in the document]
Comment (jokochems) [reply src/oemof/solph/_energy_system.py:143]: Well, this is not so very trivial. I faced some trouble with integrating the timeindex / timeincrement redefinition. Maybe, how to proceed on this is a thing to discuss at the dev meeting.

File: src/oemof/solph/_energy_system.py [hunk 4]
```diff
@@ -133,6 +161,85 @@ def __init__(
...
+                end = filter_series.loc[filter_series.index.year == year].max()
+                periods[number] = pd.date_range(start, end, freq=freq)
+        else:
+            for k in periods.keys():
+                if not isinstance(k, int):
+                    raise ValueError("Period keys must be of type int.")
+
+        return periods
+
+    def _extract_periods_years(self):
```
[the remark this answers is not in the document]
Comment (jokochems) [reply src/oemof/solph/_energy_system.py:225]: Well yeah, in pricinple it could, but it brings along some follow-up problems:
* handling discounting and annuities which is done within the model
* lifetime tracking where a mapping is made on an annual basis and periods are mapped back to "real" years.

File: src/oemof/solph/_models.py [hunk 5]
```diff
@@ -217,8 +234,8 @@ def solve(self, solver="cbc", solver_io="lp", **kwargs):
         cmdline_options : dict
             Dictionary with command line options for solver e.g.
             {"mipgap":"0.01"} results in "--mipgap 0.01"
-            {"interior":" "} results in "--interior"
-            Gurobi solver takes numeric parameter values such as
+            \{"interior":" "} results in "--interior"
+            \Gurobi solver takes numeric parameter values such as
```
[the remark this answers is not in the document]
Comment (jokochems) [reply src/oemof/solph/_models.py:238]: It serves to properly diplay a line break here if I remember correctly.

**Comment by jokochems** [comment]
Discussion outcome from dev meeting: -> Desired changes to be done before incuding:
* Standard Investments -> Change from annuities to specifying lifetime and age and implicitly calculating annuities? (Imho needs more discussion and decision making) 
* Simplification (suggestions by @gnn)
  * Let user explicitly define periods dict (mapping period number to pd.date_range objects)
  * -> Thus, get rid of attributes `multi_period=True` and `freq` which will be implicitly given
* Generalization: Mention that specifying `discount_rate=1` allows for using real values instead of nominal ones.

**Comment by jokochems** [comment]
Hi there, @oemof-developer,

in my humble opinion, the feature is ready now to become part of the code basis.

The following has changed compared to the last discussions at the dev meeting:
* `v0.5.0` structure changes have been included. All conflicts have been resolved and the tests fixed resp. updated.
* `energy_system` has been simplified. The `multi_period` attribute has been cut in favour of demanding for an explicit specification of periods, as suggested by @gnn 
* User guide has been extended and adjusted and also docs have been revised.

Please feel free to add your review.
If I don't hear anything within the next two weeks, i.e. until Friday, 12th, I'll go ahead and merge.
I think we can still add fixes once it has been integrated into dev.

There still persists the coveralls issue, but I assume this has nothing to do with the changes I introduce here. I'd be happy to see all checks succeeding, though.

**Comment by jokochems** [comment]
To simplify the review, I updated the presentation giving an overview over the feature. Please find it here: https://github.com/jokochems/multiperiod_oemof-solph/blob/main/multi-period_modeling_oemof-solph_overview.pdf

**Review by p-snft — CHANGES_REQUESTED** [review]
Honestly, I did not read through all the extra indexes you probably added. Here, there are just two tiny remarks. But please do also have a look at #935, where I already implemented what I would have as a feedback.

File: src/oemof/solph/_energy_system.py [hunk 6]
```diff
@@ -20,6 +21,37 @@
 import numpy as np
 import pandas as pd
 from oemof.network import energy_system as es
+from oemof.tools import debugging
+
+
+def _add_periods(periods):
```
[the remark this answers is not in the document]
Comment (jokochems) [reply src/oemof/solph/_energy_system.py:27]: Yes, you are right. It is a check, but used to be something different. I forgot to adjust the naming. Nevermind, since your improvements from #935 make it obsolete anyways.

File: docs/whatsnew/v0-5-0.rst [hunk 7]
```diff
@@ -16,8 +16,8 @@ API changes
 New features
 ############
 
-* Add `inactivity_costs` as an option for `Flow`s. Inactivity costs is a cost for times where a Flow is not operated.
-* It is now possible to optimise the size of `Flow`s that have a status variable. Internally, this is done by the `InvestNonConvexFlowBlock` which provides the possibility to perform capacity optimization of assets considering their min/max loads, as well as their operation status.
+* Add `inactivity_costs` as an option for `Flow`\s. Inactivity costs is a cost for times where a Flow is not operated.
+* It is now possible to optimise the size of `Flow`\s that have a status variable. Internally, this is done by the `InvestNonConvexFlowBlock` which provides the possibility to perform capacity optimization of assets considering their min/max loads, as well as their operation status.
```
[the remark this answers is not in the document]
Comment (jokochems) [reply docs/whatsnew/v0-5-0.rst:20]: I experienced a warning when I tried to render the docs: "WARNING: Inline interpreted text or phrase reference start-string without end-string.". The backslash was a solution I found online: https://stackoverflow.com/questions/58633196/restructuredtext-inline-code-followed-by-s-results-in-inline-interpreted-text, but reading through it again, it seems to be discouraged.

If I remember correctly, even the respective test failed (I have no working tox setup currently, but tried using the CI).

Long story short: It was rather a quick fix. I would be fine to remove it again in case the tests do allow to do so.

[the remark this answers is not in the document]
Comment (p-snft) [reply docs/whatsnew/v0-5-0.rst:20]: As "\s" is sometimes read as space, I really want to avoid those. I'd prefer a workaround like in:
> `Flow` objects

[the remark this answers is not in the document]
Comment (jokochems) [reply docs/whatsnew/v0-5-0.rst]: Yes, thank you. I agree with that and introduced the suggested workaround.

**Review by jokochems — COMMENTED** [review]
Thank you for the improvement and suggestions, @p-snft. I've updated the usage example and the what's new accordingly. Now your requested changes should be resolved, aren't they? Thanks once again! :-)

**Review by p-snft — APPROVED** [review]
Looks good now. (We will probably find things to improve while using it.)

**Comment by jokochems** [comment]
@p-snft Thank you very much. I totally agree that there'll likely be gradual improvements in the future.
@oemof-developer would be happy to receive another review. If I don't get any objection, I'll merge on May, 12th.

**Comment by jokochems** [comment]
There have been no objections, so I'll finally merge this pull request.

Thank you, @oemof-developer for all the fruitful discussions we had and the valuable hints.
A special thank you goes to @p-snft for the review. I know it is tedious to go through the large diffs. Also I think, there still will be gradual improvements building on the pull request.

I'm happy to assist in case this brings conflicts with some ongoing other development.
